node/src/components/event_stream_server/tests.rs

```diff
@@ -0,0 +1,899 @@
+            EventFilter::Signatures => SSE_API_SIGNATURES_PATH,
+        },
+        match maybe_start_from {
+            Some(start_from) => format!("?{}={}", QUERY_FIELD, start_from),
+            None => String::new(),
+        }
+    )
+}
+
+/// The representation of an SSE event as received by a subscribed client.
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct ReceivedEvent {
+    id: Option<Id>,
+    data: String,
+}
+
+/// Runs a client, consuming all SSE events until the server starts emitting keepalive chars `:`.
+///
+/// The client waits at the barrier before connecting to the server, and then again immediately
+/// after connecting to ensure the server doesn't start sending events before the client is
+/// connected.
+async fn subscribe(url: &str, barrier: Arc<Barrier>) -> Result<Vec<ReceivedEvent>, reqwest::Error> {
+    let client_id = CLIENT_ID.with(|client_id| {
+        let mut id = client_id.borrow_mut();
+        let copied_id = *id;
+        *id = copied_id + 1;
+        format!("client {}", copied_id)
+    });
+
+    debug!("{} waiting before connecting via {}", client_id, url);
+    barrier.wait().await;
+    let response = reqwest::get(url).await?;
+    debug!("{} waiting after connecting", client_id);
+    barrier.wait().await;
+    debug!("{} finished waiting", client_id);
+
+    // The stream from the server is not always chunked into events, so gather the stream into a
+    // single `String` until we receive a keepalive.
+    let mut response_text = String::new();
+    let mut stream = response.bytes_stream();
+    while let Some(item) = stream.next().await {
+        // If the server crashes or returns an error in the stream, it is caught here as `item` will
+        // be an `Err`.
+        let bytes = item?;
+        let chunk = str::from_utf8(bytes.as_ref()).unwrap();
+        response_text.push_str(chunk);
+        if response_text.lines().rev().any(|line| line == ":") {
+            debug!("{} received keepalive: exiting", client_id);
+            break;
+        }
+    }
+
+    // Iterate the lines of the response body.  Each line should be one of
+    //   * an SSE event: line starts with "data:" and the remainder of the line is a JSON object
+    //   * an SSE event ID: line starts with "id:" and the remainder is a decimal encoded `u32`
+    //   * empty
+    //   * a keepalive: line contains exactly ":"
+    //
+    // The expected order is:
+    //   * data:<JSON-encoded ApiVersion> (note, no ID line follows this first event)
+    // then the following three repeated for as many events as are applicable to that stream:
+    //   * data:<JSON-encoded event>
+    //   * id:<integer>
+    //   * empty line
+    // then finally, repeated keepalive lines until the server is shut down.
+    let mut received_events = Vec::new();
+    let mut line_itr = response_text.lines();
+    while let Some(data_line) = line_itr.next() {
+        let data = match data_line.strip_prefix("data:") {
```

Comment:
Maybe I've written a few too many parsers in my recent time, but how about adding a function

`fn parse_line<T: FromStr>(expected_key: &str, value: T) -> Result<T, ParseError>` (or a custom trait instead of `FromStr`)

and then writing this more readable as something roughly along the lines of 

```
let data: ApiVersion = parse_line("data", line_itr.next().ok_or(ParseError::ExpectedLine)?)?;
let data: Event = parse_line("data, line_itr.next().ok_or(ParseError::ExpectedLine)?)?;
let data: i64 = parse_line("id, line_itr.next().ok_or(ParseError::ExpectedLine)?)?;
// ...
```

Reply:
Hmm - that seems overly complex for what we need here.  While it would simplify the call site a little, it wouldn't be as simple as you show.  For example, parsing the data line is as simple as stripping the prefix `data:`.  However, we also handle the case where the expected data line is empty or is a keepalive.

I'll move the parsing wholesale to a new function as you suggest in the comment above to see if that helps a little.

Reply:
Hmm - that seems overly complex for what we need here.  While it would simplify the call site a little, it wouldn't be as simple as you show.  For example, parsing the data line is as simple as stripping the prefix `data:`.  However, we also handle the case where the expected data line is empty or is a keepalive.

I'll move the parsing wholesale to a new function as you suggest in the comment above to see if that helps a little.

Reply:
Sounds good to me. In the end, we probably want to have a proper SSE parser and not some quick inline solution I guess though.